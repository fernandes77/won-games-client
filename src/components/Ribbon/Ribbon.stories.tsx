import type { Meta, StoryObj } from '@storybook/react'
import { Ribbon } from './Ribbon'

const meta: Meta<typeof Ribbon> = {
  component: Ribbon
}

export default meta
type Story = StoryObj<typeof Ribbon>

export const Primary: Story = {
  args: { children: 'Best Seller' },
  decorators: [
    (Story) => (
      <div
        style={{
          width: '40rem',
          height: '25rem',
          position: 'relative',
          backgroundColor: '#888'
        }}
      >
        <Story />
      </div>
    )
  ]
}
