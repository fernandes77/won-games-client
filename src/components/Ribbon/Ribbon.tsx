import { Children } from '@/types/children'
import { ClassName } from '@/types/class-name'
import { cn } from '@/utils/merge-classnames'

export type RibbonColors = 'primary' | 'secondary'
export type RibbonSizes = 'md' | 'sm'

type RibbonProps = {
  color?: 'primary' | 'secondary'
  size?: 'md' | 'sm'
} & Children &
  ClassName

export const ribbonSizeMap = {
  md: 'text-sm h-9 px-sm right-[-20px] before:top-9 before:border-t-[10px] before:border-r-[20px]',
  sm: 'text-xs h-[1.625rem] px-xs right-[-15px] before:top-[26px] before:border-t-[7px] before:border-r-[15px]'
}
const colorMap = {
  primary:
    'bg-primary before:border-t-primary-dark before:border-l-primary-dark',
  secondary:
    'bg-secondary before:border-t-secondary-dark before:border-l-secondary-dark'
}

export const Ribbon = ({
  children,
  color = 'primary',
  size = 'md',
  className
}: RibbonProps) => (
  <div
    className={cn(
      `${ribbonSizeMap[size]} ${colorMap[color]}
      absolute top-xs flex items-center font-bold text-white
      before:absolute before:right-0 before:border-solid before:border-l-0
      before:border-r-transparent before:border-b-transparent before:border-b-[0.625rem]
      `,
      className
    )}
  >
    {children}
  </div>
)
