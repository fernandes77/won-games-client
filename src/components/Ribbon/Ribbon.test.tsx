import { render, screen } from '@/utils/tests/helpers'
import { Ribbon, ribbonSizeMap } from './Ribbon'

describe('<Ribbon />', () => {
  it('should render the text correctly', () => {
    const { container } = render(<Ribbon>Best Seller</Ribbon>)

    expect(screen.getByText(/Best Seller/i)).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render with the primary color', () => {
    render(<Ribbon>Best Seller</Ribbon>)

    expect(screen.getByText(/Best Seller/i)).toHaveClass('bg-primary')
  })

  it('should render with the secondary color', () => {
    render(<Ribbon color="secondary">Best Seller</Ribbon>)

    expect(screen.getByText(/Best Seller/i)).toHaveClass('bg-secondary')
  })

  it('should render with the normal size as default', () => {
    render(<Ribbon>Best Seller</Ribbon>)

    expect(screen.getByText(/Best Seller/i)).toHaveClass(ribbonSizeMap.md)
  })

  it('should render with the small size', () => {
    render(<Ribbon size="sm">Best Seller</Ribbon>)

    expect(screen.getByText(/Best Seller/i)).toHaveClass(ribbonSizeMap.sm)
  })
})
