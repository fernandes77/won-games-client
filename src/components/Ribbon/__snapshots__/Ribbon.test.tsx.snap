// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Ribbon /> should render the text correctly 1`] = `
<div
  class="text-sm h-9 px-sm right-[-20px] before:top-9 before:border-t-[10px] before:border-r-[20px] bg-primary before:border-t-primary-dark before:border-l-primary-dark absolute top-xs flex items-center font-bold text-white before:absolute before:right-0 before:border-solid before:border-l-0 before:border-r-transparent before:border-b-transparent before:border-b-[0.625rem]"
>
  Best Seller
</div>
`;
